rir/src/interpreter/interp.cpp

```diff
@@ -491,6 +491,264 @@ RIR_INLINE SEXP legacyCall(const CallContext& call, Context* ctx) {
+                return nullptr;
+
+            if (thistype == INTSXP && type == LGLSXP)
+                type = INTSXP;
+
+            if (thistype == REALSXP && type != REALSXP)
+                type = REALSXP;
+
+            total += XLENGTH(args[i]);
+        }
+
+        if (total == 0)
+            return nullptr;
+
+        long pos = 0;
+        auto res = Rf_allocVector(type, total);
+        for (size_t i = 0; i < nargs; ++i) {
+            auto len = XLENGTH(args[i]);
+            for (long j = 0; j < len; ++j) {
+                assert(pos < total);
+                if (type == REALSXP) {
+                    if (TYPEOF(args[i]) == REALSXP) {
+                        REAL(res)[pos++] = REAL(args[i])[j];
+                    } else {
+                        SLOWASSERT(NA_INTEGER == NA_LOGICAL);
+                        if (INTEGER(args[i])[j] == NA_INTEGER) {
+                            REAL(res)[pos++] = NA_REAL;
+                        } else {
+                            REAL(res)[pos++] = INTEGER(args[i])[j];
+                        }
+                    }
+                } else {
+                    INTEGER(res)[pos++] = INTEGER(args[i])[j];
+                }
+            }
+        }
+        return res;
+    }
+
+    case 109: { //"vector"
+        if (nargs != 2)
+            return nullptr;
+        if (TYPEOF(args[0]) != STRSXP)
+            return nullptr;
+        if (XLENGTH(args[0]) != 1)
+            return nullptr;
+        size_t length = -1;
+        if (IS_SIMPLE_SCALAR(args[1], INTSXP)) {
+            if (INTEGER(args[1])[0] == NA_INTEGER)
+                return nullptr;
+            length = INTEGER(args[1])[0];
+        } else if (IS_SIMPLE_SCALAR(args[1], REALSXP)) {
+            if (REAL(args[1])[0] == NA_REAL)
+                return nullptr;
+            length = REAL(args[1])[0];
+        } else {
+            return nullptr;
+        }
+        int type = str2type(CHAR(args[0]));
+
+        switch (type) {
+        case LGLSXP:
+        case INTSXP: {
+            auto res = allocVector(type, length);
+            Memzero(INTEGER(res), length);
+            return res;
+        }
+        case CPLXSXP: {
+            auto res = allocVector(type, length);
+            Memzero(COMPLEX(res), length);
+            return res;
+        }
+        case RAWSXP: {
+            auto res = allocVector(type, length);
+            Memzero(RAW(res), length);
+            return res;
+        }
+        case REALSXP: {
+            auto res = allocVector(type, length);
+            Memzero(REAL(res), length);
+            return res;
+        }
+        case STRSXP:
+        case EXPRSXP:
+        case VECSXP:
+            return allocVector(type, length);
+        case LISTSXP:
+            if (length > INT_MAX)
+                return nullptr;
+            return allocList((int)length);
+        default:
+            return nullptr;
+        }
+        assert(false);
```

Comment:
Is this `assert` necessary?

Reply:
yeah, if you forget to return or break above, it will fall through. Those are really annoying bugs, that burned me in the past. putting an asert there, is just for me to remember that I wanted to handle all the cases above, should I ever refactor that part.